File: app/src/main/java/com/mezcode/demo/snap2contact/ocrreader/SetContactFieldsActivity.java
package com.mezcode.demo.snap2contact.ocrreader;

import android.content.ContentValues;
import android.content.Intent;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.provider.ContactsContract.Contacts;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Spinner;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.HashMap;

/**
 * Created by emezias on 4/20/17.
 * This class takes the scanned text as an extra
 * It displays each line of text in its own edit text
 * The ContactSpinnerAdapter getIndices code applies logic to set the correct spinner value
 */

public class SetContactFieldsActivity extends AppCompatActivity implements AdapterView.OnItemSelectedListener {

    public static final String TAG = SetContactFieldsActivity.class.getSimpleName();
    String[] mContactFields;
    int[] editFields = new int[] { R.id.cc_edit1, R.id.cc_edit2, R.id.cc_edit3, R.id.cc_edit4, R.id.cc_edit5,
            R.id.cc_edit6, R.id.cc_edit7, R.id.cc_edit8, R.id.cc_edit9, R.id.cc_edit10 };
    int[] spinners = new int[] { R.id.cc_spinner1, R.id.cc_spinner2, R.id.cc_spinner3, R.id.cc_spinner4, R.id.cc_spinner5,
            R.id.cc_spinner6, R.id.cc_spinner7, R.id.cc_spinner8, R.id.cc_spinner9, R.id.cc_spinner10 };
    int[] btn_fields = new int[] { R.id.cc_btn1, R.id.cc_btn2, R.id.cc_btn3, R.id.cc_btn4, R.id.cc_btn5,
            R.id.cc_btn6, R.id.cc_btn7, R.id.cc_btn8, R.id.cc_btn9, R.id.cc_btn10 };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_contact);
    }

    /**
     * The field setup is done in onResume to always show the data, even if the user presses back
     */
    @Override
    protected void onResume() {
        super.onResume();
        //These extras are passed into the activity
        mContactFields = getIntent().getExtras().getStringArray(TAG); //scanned text
        final int[] spinDexes = ContactSpinnerAdapter.getIndices(mContactFields);
        //spinner index integers to set different labels for all rows

        ContactSpinnerAdapter adapter;
        Spinner spinner;
        EditText text;
        ImageButton btn;
        String s;
        for (int i = 0; i < mContactFields.length; i++) {
            s = mContactFields[i];
            adapter = new ContactSpinnerAdapter(this);
            text = (EditText) findViewById(editFields[i]);
            btn = (ImageButton) findViewById(btn_fields[i]);
            text.addTextChangedListener(new EditButton(btn));
            spinner = (Spinner)findViewById(spinners[i]);
            spinner.setAdapter(adapter);
            if (!TextUtils.isEmpty(s)) {
                text.setText(s);
                if (i < spinDexes.length && spinDexes[i] > 0) {
                    spinner.setSelection(spinDexes[i]);
                } else {
                    spinner.setSelection(i);
                }
            } else {
                spinner.setSelection(i);
                btn.setEnabled(false);
            }
            spinner.setTag(i);
            btn.setTag(i);
            //important to set listener
            spinner.setOnItemSelectedListener(this);
        } //end for loop

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.save_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        super.onOptionsItemSelected(item);
        saveContact(null);
        return true;
    }

    /**
     * This method is set in the view xml and called by the action view on the app bar
     * It alerts the user if more than one text value is mapped to the same contact column
     * If not, it calls create contact to put together the Contacts content provider intent and extras
     * @param v
     */
    public void saveContact(View v) {
        Log.i(TAG, "open Contacts intent");
        ArrayList<Integer> duplicates = new ArrayList<>();
        HashMap<Integer, String> contactMap = new HashMap<>();
        String value;
        int spinnerDex;
        for(int dex = 0; dex < editFields.length; dex++) {
            value = ((EditText) findViewById(editFields[dex])).getText().toString();
            if (!TextUtils.isEmpty(value)) {
                spinnerDex = ((Spinner) findViewById(spinners[dex])).getSelectedItemPosition();
                if (contactMap.get(spinnerDex) != null) {
                    duplicates.add(editFields[dex]);
                } else {
                    contactMap.put(spinnerDex, value);
                }
            }
        }
        if (duplicates.isEmpty()) {
            createContact(contactMap);
        } else showDuplicatesDialog(duplicates, contactMap);
    }

    //This demo code comes from the ContactsContract documentation
    //https://developer.android.com/reference/android/provider/ContactsContract.Intents.Insert.html
    void createContactTest() {
        ArrayList<ContentValues> data = new ArrayList<ContentValues>();

        ContentValues row1 = new ContentValues();
        row1.put(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Organization.CONTENT_ITEM_TYPE);
        row1.put(ContactsContract.CommonDataKinds.Organization.COMPANY, "Android");
        data.add(row1);

        ContentValues row2 = new ContentValues();
        row2.put(Contacts.Data.MIMETYPE, ContactsContract.CommonDataKinds.Email.CONTENT_ITEM_TYPE);
        row2.put(ContactsContract.CommonDataKinds.Email.TYPE, ContactsContract.CommonDataKinds.Email.TYPE_CUSTOM);
        row2.put(ContactsContract.CommonDataKinds.Email.LABEL, "Green Bot");
        row2.put(ContactsContract.CommonDataKinds.Email.ADDRESS, "android@android.com");
        data.add(row2);

        Intent intent = new Intent(Intent.ACTION_INSERT, Contacts.CONTENT_URI);
        intent.putParcelableArrayListExtra(ContactsContract.Intents.Insert.DATA, data);

        startActivity(intent);
    }

    /******** Click listeners *********/

    public void clearText(View btn) {
        btn.setEnabled(false);
        ((EditText)findViewById(editFields[(Integer) btn.getTag()])).getText().clear();
    }

    private void createContact(HashMap<Integer, String> contactMap) {
        // Creates a new Intent to insert a contact
        ArrayList<ContentValues> data = new ArrayList<>();
        ContentValues contactdata; //hint
        //each content value fills a contacts contract row, the name is set as an extra
        Intent intent = new Intent(Intent.ACTION_INSERT, Contacts.CONTENT_URI);
        for (int key: contactMap.keySet()) {
            switch(key) {
                case ContactSpinnerAdapter.IND_NAME:
                    //could not find the right combination of content value keys for the name
                    intent.putExtra(ContactsContract.Intents.Insert.NAME, contactMap.get(key));
                    break;
                case ContactSpinnerAdapter.IND_PHONE:
                    //TODO
                    break;
                case ContactSpinnerAdapter.IND_EMAIL:
                    //TODO
                    break;
                case ContactSpinnerAdapter.IND_ADDR:
                    //TODO
                    break;
                case ContactSpinnerAdapter.IND_COMPANY:
                    //TODO
                case ContactSpinnerAdapter.IND_TITLE:
                    //TODO
                    break;
                case ContactSpinnerAdapter.IND_IM:
                    //TODO
                    break;
                case ContactSpinnerAdapter.IND_NOTES:
                    //TODO
                    break;
                case ContactSpinnerAdapter.IND_PHONE2:
                    //TODO
                    break;
                case ContactSpinnerAdapter.IND_EMAIL2:
                    //TODO
                    break;
            }
        }

        intent.putParcelableArrayListExtra(ContactsContract.Intents.Insert.DATA, data);
        startActivity(intent);

        Toast.makeText(this, "Creating new contact: " + intent.getStringExtra(ContactsContract.Intents.Insert.NAME), Toast.LENGTH_LONG).show();
    }

    void showDuplicatesDialog(ArrayList<Integer> duplicates, final HashMap<Integer, String> contactMap) {
        //TODO
        //Sample code uses Alert Dialog builder instead of Dialog Fragment
    }

    //The tag set on the spinner view here is used to set a contact field label when the user saves
    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        final int previous = (int) parent.getTag();
        if (position == previous) return;
        parent.setTag(position);
        //Move map to Contact Fields logic check to save button, nothing to do here!
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) { }
}

File: app/src/main/java/com/mezcode/demo/roloscan/ocrreader/SetContactFieldsActivity.kt
package com.mezcode.demo.roloscan.ocrreader

import android.content.Intent
import android.os.Bundle
import android.provider.ContactsContract
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.mezcode.demo.roloscan.ocrreader.databinding.ActivityCreateContactBinding
import dagger.hilt.android.AndroidEntryPoint

/**
 * Created by emezias on 4/20/17.
 * This class takes the scanned text as an extra
 * It displays each line of text in its own edit text
 * The ContactSpinnerAdapter getIndices code applies logic to set the correct spinner value
 * TBH this logic isn't very useful - it is intended to demonstrate the ContactsContract
 * Need to update to use 'query' params in the Manifest
 */
@AndroidEntryPoint
class SetContactFieldsActivity : AppCompatActivity() {
    
    private lateinit var adapter: ScannedTextAdapter
    private lateinit var binding: ActivityCreateContactBinding

    companion object {
        val TAG: String = this::class.java.simpleName
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCreateContactBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val contactFields = (intent.extras?.getStringArray(TAG) as Array<String>).toMutableList()
        //scanned text from StartActivity view model instance is packed as an extra
        //could put a different view model here
        if (contactFields.isEmpty()) {
            Utils.showSnackbar(binding.progress, R.string.no_text)
            finish()
            return
        }
        adapter = ScannedTextAdapter(contactFields)
        binding.contactFields.adapter = adapter
        binding.contactFields.layoutManager = LinearLayoutManager(this@SetContactFieldsActivity)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.save_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        super.onOptionsItemSelected(item)
        saveContact()
        return true
    }


    /**
     * This method is called by the action view on the app bar
     * It alerts the user if no text value is mapped to the name contact column
     * Otherwise it calls create contact to put together the Contacts content provider intent and extras
     */
    private fun saveContact() {
        Log.i(TAG, "open Contacts intent")
        val scannedTextList = adapter.returnContactFields()
        val nameFields =
            scannedTextList.filter { it.contactLabel == Utils.SpinnerIndex.IND_NAME }
        if (nameFields.isEmpty() || nameFields.size > 1 ) {
            Utils.showSnackbar(binding.progress, R.string.contact_error)
            return
        }

        // The contact contract should have one and only one name field, let's see about combining
        val intent = Intent(ContactsContract.Intents.Insert.ACTION).apply {
            // Sets the MIME type to match the Contacts Provider
            type = ContactsContract.RawContacts.CONTENT_TYPE
            for (contactExtra in scannedTextList) {
                if (contactExtra.contactLabel == null) {
                    putExtra(ContactsContract.Intents.Insert.NOTES, contactExtra.scannedText)
                } else {
                    putExtra(
                        contactExtra.contactLabel?.key ?: ContactsContract.Intents.Insert.NOTES,
                        contactExtra.scannedText
                    )
                }
            }
        } 
        //intent creation complete - boots and suspenders check for a name
        if (intent.hasExtra(ContactsContract.Intents.Insert.NAME)) {
            if (intent.hasExtra(ContactsContract.Intents.Insert.PHONE)) {
                intent.putExtra(ContactsContract.Intents.Insert.PHONE_ISPRIMARY, true)
            }
            startActivity(intent)
            finish()
            //end the activity, don't want to create the same contact twice
        } else {
            Utils.showSnackbar(binding.progress, R.string.contact_error)
            return
        }

    }

    //This logic is guided by the ContactsProvider documentation on the Android dev site
    //https://developer.android.com/training/contacts-provider/modify-data

}